### PR TITLE
Reset and and parametric gates in `kernel_builder`

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

- Use `const QuakeValue &` for `reset` rather than `QuakeValue &` to handle the `reset(q_reg[idx])` use case since `QuakeValue::operator[]` returns a `QuakeValue` rval.

- `CUDAQ_BUILDER_ADD_ONE_QUBIT_PARAM` to handle floating-point params in adjoint and control code paths (in addition to the base one) by converting the floating point number to a `QuakeValue` if necessary.

### CODE REVIEW

File: runtime/cudaq/builder/kernel_builder.h [hunk 1]
```diff
@@ -444,24 +453,27 @@ class kernel_builder : public details::kernel_builder_base {
     QuakeValue v(*opBuilder, param);                                           \
     details::NAME(*opBuilder, v, empty, qubit);                                \
   }                                                                            \
-  template <NumericType ConstantParameter>                                     \
-  void NAME(ConstantParameter &&constant, QuakeValue qubit) {                  \
-    NAME(constant, qubit);                                                     \
```
Comment (1tnguyen) [inline runtime/cudaq/builder/kernel_builder.h:449]: This overload caused an infinite recursive warning when the constant is a `float`, for example. The `void NAME(double param, QuakeValue qubit)` overload seems to work for all floating point types.

File: runtime/cudaq/builder/kernel_builder.h [hunk 2]
```diff
@@ -444,24 +453,27 @@ class kernel_builder : public details::kernel_builder_base {
...
-  }                                                                            \
-  template <typename mod,                                                      \
+  template <typename mod, QuakeValueOrNumericType paramT,                      \
```
Comment (amccaskey) [inline runtime/cudaq/builder/kernel_builder.h:456]: Would you mind capitalizing this type - ParamT

File: runtime/cudaq/builder/kernel_builder.h [hunk 3]
```diff
@@ -444,24 +453,27 @@ class kernel_builder : public details::kernel_builder_base {
...
+      NAME(QuakeValue(*opBuilder, -parameter), qubit);                         \
+    else                                                                       \
+      NAME(-parameter, qubit);                                                 \
   }                                                                            \
-  template <typename mod, typename... QubitValues,                             \
+  template <typename mod, QuakeValueOrNumericType paramT,                      \
+            typename... QubitValues,                                           \
             typename = typename std::enable_if_t<sizeof...(QubitValues) >= 2>> \
-  void NAME(QuakeValue parameter, QubitValues... args) {                       \
+  void NAME(paramT &&parameter, QubitValues... args) {                         \
```
Comment (amccaskey) [inline runtime/cudaq/builder/kernel_builder.h:468]: Should we just go ahead and update this to `const ParamT&` and the one above it?

**Comment by amccaskey** [comment]
Thanks @1tnguyen I think I had a similar fix in one of my branches that I hadn't PR'ed yet. You beat me to it :)

File: runtime/cudaq/builder/kernel_builder.h (hunk 2 above)
Comment (1tnguyen) [reply runtime/cudaq/builder/kernel_builder.h]: Definitely. Fixed.

File: runtime/cudaq/builder/kernel_builder.h (hunk 3 above)
Comment (1tnguyen) [reply runtime/cudaq/builder/kernel_builder.h]: It's a good idea. Changed both to `const ParamT&`.

**Review by amccaskey — APPROVED** [review]

**Pull request merged**